clients/src/main/java/org/apache/kafka/clients/admin/internals/DeleteRecordsHandler.java

```diff
@@ -0,0 +1,170 @@
+                    .setOffset(entry.getValue().beforeOffset()));
+
+            System.out.println("Partitions: " + deleteRecords.partitions());
+        }
+
+        DeleteRecordsRequestData data = new DeleteRecordsRequestData()
+                .setTopics(new ArrayList<>(deletionsForTopic.values()));
+        return new DeleteRecordsRequest.Builder(data);
+    }
+
+
+    @Override
+    public ApiResult<TopicPartition, DeletedRecords> handleResponse(
+        Node broker,
+        Set<TopicPartition> keys,
+        AbstractResponse abstractResponse
+    ) {
+        DeleteRecordsResponse response = (DeleteRecordsResponse) abstractResponse;
+        Map<TopicPartition, DeletedRecords> completed = new HashMap<>();
+        Map<TopicPartition, Throwable> failed = new HashMap<>();
+        List<TopicPartition> unmapped = new ArrayList<>();
+        Set<TopicPartition> retriable = new HashSet<>();
+
+        for (DeleteRecordsResponseData.DeleteRecordsTopicResult topicResult: response.data().topics()) {
+            for (DeleteRecordsResponseData.DeleteRecordsPartitionResult partitionResult : topicResult.partitions()) {
+                Errors error = Errors.forCode(partitionResult.errorCode());
+                TopicPartition topicPartition = new TopicPartition(topicResult.name(), partitionResult.partitionIndex());
+                if (error == Errors.NONE) {
+                    completed.put(topicPartition, new DeletedRecords(partitionResult.lowWatermark()));
+                } else {
+                    handlePartitionError(topicPartition, error, failed, unmapped, retriable);
+                }
+            }
+        }
+
+        // Sanity-check if the current leader for these partitions returned results for all of them
+        for (TopicPartition topicPartition : keys) {
+            if (unmapped.isEmpty()
+                    && !completed.containsKey(topicPartition)
+                    && !failed.containsKey(topicPartition)
+                    && !retriable.contains(topicPartition)
+            ) {
+                ApiException sanityCheckException = new ApiException(
+                        "The response from broker " + broker.id() +
+                                " did not contain a result for topic partition " + topicPartition);
+                log.error(
+                        "DeleteRecords request for topic partition {} failed sanity check",
+                        topicPartition,
+                        sanityCheckException);
+                failed.put(topicPartition, sanityCheckException);
+            }
+        }
+
+        return new ApiResult<>(completed, failed, unmapped);
+    }
+
+    private void handlePartitionError(
+        TopicPartition topicPartition,
+        Errors error,
+        Map<TopicPartition, Throwable> failed,
+        List<TopicPartition> unmapped,
+        Set<TopicPartition> retriable
+    ) {
+        if (error == Errors.NOT_LEADER_OR_FOLLOWER || error == Errors.LEADER_NOT_AVAILABLE) {
```

Comment:
Please extract out handling of NOT_LEADER_OR_FOLLOWER at a common place for clients: e.g. other usage https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java#L158